File: GoEngineering/03_api_client_with_retries/explanation/03_api_client_with_retries_explanation.go
/*
Problem 03 — API Client With Retries (backoff, jitter, circuit breaker, deadlines)

WHAT WE'RE BUILDING

An HTTP client wrapper (`Client`) that calls a flaky upstream service
resiliently: retries transient failures with exponential backoff and
jitter, respects context deadlines/cancellation at every step, refuses to
retry non-idempotent-unsafe or client-error (4xx) responses, and trips a
circuit breaker after sustained failures so a downed dependency doesn't
get hammered by every caller's individual retry loop.

# WHY THIS MATTERS IN REAL SYSTEMS

Every service that calls another service over the network needs this
exact logic, and getting it wrong is one of the most common causes of
cascading outages: naive retries without backoff turn a brief blip into a
"retry storm" that keeps the downstream dead; retries without jitter
synchronize every client's retry attempts into thundering-herd waves;
retries without a circuit breaker mean a fully-down dependency still gets
hit by 100% of traffic, wasting connections/threads/latency budget on
calls that are certain to fail; and retries that ignore context deadlines
keep working past the point where the caller has already given up and
walked away.

CONCEPTS COVERED

  - Exponential backoff with jitter (full jitter, per Marc Brooker's
    well-known formula: `sleep = random(0, min(cap, base*2^attempt))`)
  - Retry classification: which errors/status codes are worth retrying
    (network errors, 429, 5xx) vs not (4xx other than 429, context
    cancellation)
  - `context.Context` deadline propagation through retries — a deadline
    is a budget for the WHOLE operation including retries, not a
    per-attempt timeout
  - A circuit breaker state machine: Closed -> Open -> Half-Open -> Closed
  - `Retry-After` header handling (a well-behaved server tells you exactly
    how long to back off)
  - Idempotency: why only GET/PUT/DELETE (and POST specifically marked
    idempotent by the caller) should ever be retried automatically
  - Interfaces for testability: an `httpDoer` interface
    (`Do(*http.Request) (*http.Response, error)`) so tests can inject a
    fake transport without a real network call or `httptest.Server` for
    every test case

# SPEC

Build a `Client` wrapping an `*http.Client` (or any `httpDoer`) with:

	type RetryPolicy struct {
	    MaxAttempts int           // total attempts including the first, >= 1
	    BaseDelay   time.Duration // e.g. 100ms
	    MaxDelay    time.Duration // cap on any single backoff sleep
	}

	func NewClient(doer httpDoer, policy RetryPolicy) *Client

	func (c *Client) Do(ctx context.Context, req *http.Request) (*http.Response, error)

Retry rules:
  - Retry on: network-level errors (anything `doer.Do` returns as `err !=
    nil`, e.g. connection refused/reset, DNS failure — NOT context
    cancellation, which must propagate immediately without further
    attempts), HTTP 429, and HTTP 5xx (500-599).
  - Do NOT retry on: any other 4xx status (400-499 except 429), or any
    2xx/3xx (obviously — that's success).
  - Respect `Retry-After` on a 429/503 response if present (either
    delta-seconds or an HTTP-date; delta-seconds is enough for this
    problem — document the HTTP-date gap as a known limitation if you
    skip it) — use it INSTEAD OF the computed backoff delay for that
    attempt, but still cap it to a sane maximum so a malicious/broken
    server can't make you sleep for a day.
  - Stop retrying after MaxAttempts total attempts (so MaxAttempts=3 means
    at most 2 retries) OR when ctx is done, whichever comes first.
  - Between attempts, sleep for `min(MaxDelay, BaseDelay * 2^(attempt-1))`
    with full jitter applied (uniform random in [0, that value]), UNLESS
    Retry-After overrides it. The sleep must itself respect ctx
    cancellation (a `time.Sleep` cannot be interrupted — you need a
    `select` on a timer and `ctx.Done()`).
  - On exhausting all attempts, return the last response/error, wrapped
    with context (e.g. `fmt.Errorf("after %d attempts: %w", n, err)`), NOT
    a generic "retries exhausted" with no trace of the underlying cause.

Circuit breaker (`CircuitBreaker`, wrapping a `Client` or standalone,
your call — recommended: standalone, composed via a wrapping RoundTripper
or a decorator around `httpDoer`):
  - States: Closed (normal), Open (fail fast, no calls attempted), Half-
    Open (one trial call allowed through to test recovery).
  - Trips to Open after N consecutive failures (configurable).
  - After a configurable cooldown, transitions Open -> Half-Open
    automatically (time-based, no external trigger needed).
  - A successful Half-Open trial call closes the circuit (reset failure
    count); a failed one reopens it (reset cooldown timer).
  - Must be safe for concurrent use (many goroutines calling through the
    same breaker).
  - When Open, calls fail immediately with a sentinel `ErrCircuitOpen`
    without attempting any network I/O or consuming a retry budget.

ACCEPTANCE CRITERIA

  - A fake `httpDoer` that fails N times then succeeds proves the client
    retries exactly enough times and no more.
  - A fake that always returns 400 proves the client does NOT retry
    client errors.
  - A fake honoring `context.Canceled` context proves retries stop
    immediately rather than continuing to sleep/retry.
  - A backoff-only test (no real sleeping — inject a clock/sleep function)
    proves delays grow exponentially and are capped.
  - A circuit breaker test proves: N consecutive failures trips it,
    subsequent calls fail fast with ErrCircuitOpen without touching the
    fake doer, and after the cooldown a single trial call is let through.
  - All tests run in well under a second — no real `time.Sleep` of
    meaningful duration in tests (inject a clock or use tiny durations
    with a fake clock, not real wall time, for backoff-heavy tests).
*/
package main

import (
	"context"
	"net/http"
	"sync"
	"time"
)

// httpDoer is the minimal interface Client depends on — satisfied by
// *http.Client, and trivially fakeable in tests without a real network
// call or httptest.Server.
type httpDoer interface {
	Do(req *http.Request) (*http.Response, error)
}

// RetryPolicy configures retry attempts and backoff.
type RetryPolicy struct {
	MaxAttempts int
	BaseDelay   time.Duration
	MaxDelay    time.Duration
}

// Client wraps an httpDoer with retry/backoff logic.
//
// TODO: add whatever fields you need: the doer, the policy, and an
// injectable sleep function `func(context.Context, time.Duration) error`
// so tests can avoid real wall-clock sleeps while still exercising the
// cancellation-aware sleeping logic.
type Client struct {
	doer   httpDoer
	policy RetryPolicy
	// TODO: fields (e.g. a sleep func for testability)
}

// NewClient returns a Client wrapping doer with the given retry policy.
// TODO: implement, defaulting the sleep func to a real ctx-aware sleep.
func NewClient(doer httpDoer, policy RetryPolicy) *Client {
	panic("TODO: implement NewClient")
}

// isRetryableStatus reports whether an HTTP status code should trigger a
// retry per the spec (429 or 5xx; not other 4xx, not 2xx/3xx).
// TODO: implement.
func isRetryableStatus(status int) bool {
	panic("TODO: implement isRetryableStatus")
}

// backoffDelay computes the full-jitter exponential backoff delay for the
// given (1-indexed) attempt number, per the spec formula:
// random(0, min(MaxDelay, BaseDelay*2^(attempt-1))).
// TODO: implement. attempt=1 is the delay before the SECOND attempt (i.e.
// call this with attempt-1 already representing "how many attempts have
// been made so far" — get the indexing right and document it).
func backoffDelay(policy RetryPolicy, attempt int, rng func() float64) time.Duration {
	panic("TODO: implement backoffDelay")
}

// parseRetryAfter parses a Retry-After header value as delta-seconds and
// returns the resulting duration, or ok=false if absent/unparseable.
// TODO: implement the delta-seconds form; document HTTP-date as
// unsupported (a known, acceptable gap for this problem).
func parseRetryAfter(header string) (time.Duration, bool) {
	panic("TODO: implement parseRetryAfter")
}

// Do performs req, retrying per c.policy, and returns the first non-retryable
// response/error (or the last one after exhausting attempts).
//
// TODO: implement the full retry loop:
//  1. Loop attempt 1..MaxAttempts.
//  2. Clone req for each attempt (http.Request bodies can only be read
//     once — you need req.Body to be replayable via GetBody, or require
//     the caller supply a body that supports it; document this
//     requirement).
//  3. Call c.doer.Do. If err != nil and ctx.Err() != nil, return
//     immediately (don't retry past cancellation). If err != nil
//     otherwise, this attempt failed; retry if attempts remain.
//  4. If err == nil, check resp.StatusCode: return immediately if not
//     retryable; otherwise drain+close resp.Body (required before
//     retrying to allow connection reuse) and retry if attempts remain.
//  5. Before each retry, sleep per backoffDelay (or Retry-After if
//     present), respecting ctx cancellation during the sleep itself.
//  6. After exhausting attempts, return the last response/error wrapped
//     with attempt count context.
func (c *Client) Do(ctx context.Context, req *http.Request) (*http.Response, error) {
	panic("TODO: implement Client.Do")
}

// ---------------------------------------------------------------------------
// Circuit breaker
// ---------------------------------------------------------------------------

// CircuitState identifies which of the three circuit-breaker states a
// CircuitBreaker is in.
type CircuitState int

const (
	StateClosed CircuitState = iota
	StateOpen
	StateHalfOpen
)

// ErrCircuitOpen is returned by CircuitBreaker.Do when the circuit is
// open and no call was attempted.
var ErrCircuitOpen error

// CircuitBreaker wraps an httpDoer, tripping open after consecutive
// failures and probing recovery after a cooldown.
//
// TODO: add fields: mutex, current state, consecutive failure count,
// failure threshold, cooldown duration, time the circuit opened (to know
// when the cooldown elapses), and the wrapped doer.
type CircuitBreaker struct {
	mu sync.Mutex
	// TODO: fields
}

// NewCircuitBreaker returns a CircuitBreaker wrapping doer, tripping open
// after failureThreshold consecutive failures and attempting recovery
// after cooldown.
func NewCircuitBreaker(doer httpDoer, failureThreshold int, cooldown time.Duration) *CircuitBreaker {
	panic("TODO: implement NewCircuitBreaker")
}

// Do executes req through the breaker: fails fast with ErrCircuitOpen if
// open (and the cooldown hasn't elapsed), allows exactly one trial call
// through if half-open, and otherwise passes through to the wrapped
// doer, updating breaker state based on the outcome.
//
// TODO: implement the state machine described in the spec. Treat any
// non-nil err OR a 5xx status as a "failure" for breaker purposes (this
// is a design decision worth documenting: should 4xx count as a
// breaker failure? Spec answer: no — a 400 means the CALLER is wrong,
// not the dependency, so it shouldn't move the breaker toward Open).
func (b *CircuitBreaker) Do(req *http.Request) (*http.Response, error) {
	panic("TODO: implement CircuitBreaker.Do")
}

// State returns the breaker's current state (useful for tests/metrics).
func (b *CircuitBreaker) State() CircuitState {
	panic("TODO: implement CircuitBreaker.State")
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

func main() {
	// TODO: wire a Client around http.DefaultClient with a sane
	// RetryPolicy, optionally wrap a CircuitBreaker around
	// http.DefaultClient first and pass the breaker as the Client's
	// doer (CircuitBreaker itself implements httpDoer), then issue a
	// demo request against a well-known reliable endpoint (or a local
	// httptest.Server you spin up inline) and print the outcome —
	// demonstrating that Client -> CircuitBreaker -> *http.Client
	// composes cleanly because everything speaks httpDoer.
	panic("TODO: implement main")
}

/*
HINTS

  - Full jitter (`rand(0, cap)`) spreads retries far better than
    "equal jitter" or no jitter at all — it's the AWS Architecture Blog's
    documented recommendation and worth citing as the reasoning, not just
    the formula.
  - `req.Clone(ctx)` gives you a fresh *http.Request per attempt sharing
    the same GetBody — but you still need to call `req.GetBody()` to get
    a fresh io.ReadCloser body for each attempt if the original request
    had a body (Clone does NOT re-arm Body itself).
  - Always `io.Copy(io.Discard, resp.Body)` then `resp.Body.Close()`
    before retrying a request whose response you're discarding — failing
    to drain the body prevents the underlying connection from being
    reused by the transport's connection pool, silently degrading every
    subsequent request to a fresh TCP+TLS handshake.
  - Inject both the sleep function AND (if you want fully deterministic
    tests) the RNG function used for jitter — a test that can force
    jitter's random draw makes backoff assertions exact instead of
    range-checks.

COMMON PITFALLS

  - Retrying a POST that isn't idempotent (e.g. "charge the customer $10")
    — a network error on the response leg (the charge succeeded, but the
    ack was lost) followed by a retry double-charges. This problem's
    Client retries based on status/error only, not method — document
    that a real system additionally needs an idempotency key or must only
    auto-retry verbs the caller has confirmed are safe.
  - Treating `context.DeadlineExceeded` as a retryable network error — if
    ctx is already past its deadline, retrying just burns more time
    failing the same way; check `ctx.Err()` first.
  - `time.Sleep(d)` inside the retry loop instead of a ctx-aware sleep —
    ignores cancellation entirely, so a caller who gave up still waits
    out the full backoff.
  - A circuit breaker with no mutex — every field it tracks (state,
    failure count, last-opened time) is written from every caller's
    goroutine; without a lock this is a data race, not just a
    theoretical one (go test -race will catch it immediately).
  - Not resetting the consecutive-failure counter on ANY success — a
    breaker that only resets on reaching some other state can trip
    permanently after a transient blip even though the dependency has
    long since recovered.

STRETCH GOALS

  - Parse the HTTP-date form of Retry-After too.
  - Add per-endpoint circuit breakers (a map keyed by host) instead of one
    global breaker.
  - Emit metrics (attempt count, breaker state transitions) via a
    injectable callback/interface instead of just returning them.
  - Add a token-bucket rate limit on top using golang.org/x/time/rate
    (already vendored) to cap outbound request rate independent of
    retries.
*/

File: PyEngineering/03_api_client_with_retries/03_api_client_with_retries_explanation.py
"""
03 — API Client With Retries
=============================

WHAT
----
An async HTTP client wrapper that calls a flaky upstream service
resiliently, built on `httpx`:

    - `RetryingClient` retries transient failures (transport errors, 429,
      5xx) with capped exponential backoff and *full jitter*, honours
      `Retry-After`, refuses to retry client errors (4xx) and refuses to
      auto-retry non-idempotent requests unless the caller opted in with an
      `Idempotency-Key` header.
    - `CircuitBreaker` wraps any sender and trips to OPEN after N
      consecutive failures, so a dead dependency gets zero traffic (not
      "every caller's full retry loop") until a cooldown passes and a
      single HALF_OPEN trial call proves it has recovered.
    - Both speak the same tiny `Sender` protocol (`async send(request) ->
      response`) as `httpx.AsyncClient`, so they compose like decorators:
      `RetryingClient(CircuitBreaker(httpx.AsyncClient()))`.

WHY THIS MATTERS
-----------------
Every service that calls another service over the network needs exactly
this logic, and getting it wrong is one of the most common causes of
cascading outages:

    1. **Retries without backoff** turn a two-second blip into a retry
       storm that keeps the downstream dead. With 3 layers of services each
       retrying 3 times, one user request can become 3^3 = 27 calls at the
       bottom of the stack.
    2. **Backoff without jitter** synchronises every client: they all fail
       at t=0, all retry at t=100ms, all retry at t=300ms -- thundering-herd
       waves instead of a smooth trickle. "Full jitter"
       (`sleep = uniform(0, min(cap, base * 2**attempt))`, from the AWS
       Architecture Blog post "Exponential Backoff and Jitter") spreads
       retries across the whole window.
    3. **Retries without a circuit breaker** mean a fully-down dependency
       still gets 100% of traffic times the retry multiplier, burning
       connections and latency budget on calls that are certain to fail.
    4. **Retries that ignore the caller's deadline** keep working after the
       caller has given up. In asyncio the deadline is an
       `asyncio.timeout(...)` block around the *whole* call (all attempts
       and sleeps), and cancellation arrives as `CancelledError` -- which
       must never be caught and retried.
    5. **Retrying non-idempotent calls** double-charges customers: the POST
       succeeded, the response was lost on the way back, the retry charges
       again. Only idempotent methods (GET, HEAD, PUT, DELETE, OPTIONS) or
       requests carrying an idempotency key are safe to replay.

CONCEPTS EXERCISED
-------------------
    - Retry classification: `httpx.TransportError` (connect/read errors,
      timeouts) and 429/5xx are retryable; other 4xx are not; cancellation
      is never retried.
    - Capped exponential backoff with full jitter, with the RNG and the
      sleep function injected so tests are exact and take no wall time.
    - `Retry-After` in both forms (delta-seconds and HTTP-date, parsed with
      `email.utils.parsedate_to_datetime`), capped so a broken server
      can't make you sleep for a day.
    - Closing discarded responses (`await response.aclose()`) so the
      connection returns to the pool.
    - A circuit breaker state machine CLOSED -> OPEN -> HALF_OPEN -> CLOSED
      with an injected monotonic clock, and why asyncio's single-threaded
      model means no lock is needed as long as state changes happen between
      `await`s.
    - `typing.Protocol` for structural typing: `httpx.AsyncClient`, the
      breaker and the retrying client are all interchangeable `Sender`s.
    - Faking HTTP with `httpx.MockTransport` -- the real client stack, no
      sockets.

SPEC
----
    class Sender(Protocol):
        async def send(self, request: httpx.Request) -> httpx.Response: ...

    @dataclass(frozen=True)
    class RetryPolicy:
        max_attempts: int = 3         # total attempts incl. the first, >= 1
        base_delay: float = 0.1       # seconds
        max_delay: float = 5.0        # cap on any single computed sleep
        max_retry_after: float = 30.0 # cap on a server-supplied Retry-After

    def is_retryable_status(status: int) -> bool
        "429 or 500-599."

    def is_idempotent(request: httpx.Request) -> bool
        "GET/HEAD/OPTIONS/PUT/DELETE/TRACE, or any request that carries an
         Idempotency-Key header."

    def backoff_delay(policy, attempts_made: int, rng: Callable[[], float]) -> float
        "Delay before the next attempt after `attempts_made` (>= 1) attempts:
         rng() * min(max_delay, base_delay * 2 ** (attempts_made - 1))."

    def parse_retry_after(value: str | None, now: datetime | None = None) -> float | None
        "Seconds to wait, from delta-seconds ('120') or an HTTP-date
         ('Wed, 21 Oct 2015 07:28:00 GMT'). None if absent or unparseable.
         A date in the past means 0."

    class RetriesExhaustedError(Exception):
        attempts: int                 # raised `from` the last transport error

    class RetryingClient:
        def __init__(self, sender, policy=RetryPolicy(), *,
                     sleep=asyncio.sleep, rng=random.random) -> None
        async def send(self, request: httpx.Request) -> httpx.Response

    Retry rules for `RetryingClient.send`:
      - At most `max_attempts` attempts in total.
      - Non-idempotent requests get exactly one attempt.
      - Transport error: retry if attempts remain, else raise
        `RetriesExhaustedError(attempts)` chained `from` the last error.
      - Retryable status: close the response, retry if attempts remain,
        else *return* the last response (the caller sees the real 503 and
        its body/headers, not a generic error).
      - Anything else (2xx/3xx/other 4xx): return immediately.
      - Sleep between attempts = parsed Retry-After (capped at
        `max_retry_after`) if the response had one, else `backoff_delay`.
      - `CircuitOpenError` and `asyncio.CancelledError` propagate at once.

    class CircuitState(Enum): CLOSED, OPEN, HALF_OPEN

    class CircuitOpenError(Exception): ...

    class CircuitBreaker:
        def __init__(self, sender, *, failure_threshold: int = 5,
                     cooldown: float = 30.0, clock=time.monotonic) -> None
        @property
        def state(self) -> CircuitState   # reports HALF_OPEN once cooldown passed
        async def send(self, request: httpx.Request) -> httpx.Response

    Breaker rules:
      - A "failure" is a transport error or a 5xx response. A 4xx is the
        *caller's* fault and does not move the breaker.
      - CLOSED: pass through; `failure_threshold` consecutive failures trip
        it to OPEN. Any success resets the consecutive-failure count.
      - OPEN: raise `CircuitOpenError` without touching the wrapped sender,
        until `cooldown` seconds have passed, then become HALF_OPEN.
      - HALF_OPEN: exactly one trial request is let through; concurrent
        callers get `CircuitOpenError`. Success -> CLOSED; failure (or the
        trial being cancelled) -> OPEN with a fresh cooldown.

ACCEPTANCE CRITERIA
--------------------
    - A fake upstream that fails N times then succeeds is called exactly
      N + 1 times, and a 400 is never retried.
    - A POST without an Idempotency-Key is attempted once even on 503; the
      same POST with the header is retried.
    - Backoff grows exponentially and is capped (asserted exactly with an
      injected rng, not with range checks).
    - Retry-After overrides the computed backoff and is capped.
    - Cancelling the caller during a backoff sleep stops further attempts.
    - The breaker trips after the threshold, fails fast without calling the
      sender, lets exactly one trial through after the cooldown, and a
      failed trial reopens it.
    - The whole test file runs in well under a second (no real sleeping).

Everything below is a stub. Fill in the `# TODO:` markers. Full type hints
are already in place -- match them.
"""

from __future__ import annotations

import asyncio
import random
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Protocol

import httpx

SleepFunc = Callable[[float], Awaitable[None]]


class Sender(Protocol):
    async def send(self, request: httpx.Request) -> httpx.Response: ...


@dataclass(frozen=True)
class RetryPolicy:
    max_attempts: int = 3
    base_delay: float = 0.1
    max_delay: float = 5.0
    max_retry_after: float = 30.0

    # TODO: validate in __post_init__ (max_attempts >= 1, delays >= 0).


class RetriesExhaustedError(Exception):
    def __init__(self, attempts: int) -> None:
        super().__init__(f"request failed after {attempts} attempt(s)")
        self.attempts = attempts


class CircuitOpenError(Exception):
    pass


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


def is_retryable_status(status: int) -> bool:
    # TODO: 429 or any 5xx.
    raise NotImplementedError


def is_idempotent(request: httpx.Request) -> bool:
    # TODO: idempotent HTTP method, or an Idempotency-Key header present.
    raise NotImplementedError


def backoff_delay(
    policy: RetryPolicy, attempts_made: int, rng: Callable[[], float]
) -> float:
    # TODO: rng() * min(max_delay, base_delay * 2 ** (attempts_made - 1)).
    #  Guard against float overflow for huge attempt counts (2 ** 2000).
    raise NotImplementedError


def parse_retry_after(value: str | None, now: datetime | None = None) -> float | None:
    # TODO:
    #  - None/empty -> None
    #  - all digits -> float(value)
    #  - otherwise try email.utils.parsedate_to_datetime; seconds until that
    #    moment (clamped at 0); unparseable -> None
    raise NotImplementedError


class RetryingClient:
    def __init__(
        self,
        sender: Sender,
        policy: RetryPolicy | None = None,
        *,
        sleep: SleepFunc = asyncio.sleep,
        rng: Callable[[], float] = random.random,
    ) -> None:
        # TODO: store collaborators; default policy = RetryPolicy().
        raise NotImplementedError

    async def send(self, request: httpx.Request) -> httpx.Response:
        # TODO: the retry loop described in the SPEC.
        #  - attempts allowed = policy.max_attempts if idempotent else 1
        #  - catch httpx.TransportError only (never bare Exception: that
        #    would swallow CircuitOpenError and programming errors)
        #  - on a retryable status with attempts left: read Retry-After,
        #    `await response.aclose()`, then sleep
        #  - on exhaustion: return last response, or raise
        #    RetriesExhaustedError(n) from the last transport error
        raise NotImplementedError


class CircuitBreaker:
    def __init__(
        self,
        sender: Sender,
        *,
        failure_threshold: int = 5,
        cooldown: float = 30.0,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        # TODO: fields: sender, threshold, cooldown, clock, current state,
        #  consecutive failures, opened_at, trial_in_flight flag.
        raise NotImplementedError

    @property
    def state(self) -> CircuitState:
        # TODO: OPEN whose cooldown has elapsed reports HALF_OPEN.
        raise NotImplementedError

    async def send(self, request: httpx.Request) -> httpx.Response:
        # TODO: the state machine in the SPEC. Decide "may I call?" before
        #  the await, record the outcome after it. Use try/finally so a
        #  cancelled HALF_OPEN trial doesn't leave the breaker stuck with
        #  trial_in_flight=True forever.
        raise NotImplementedError


if __name__ == "__main__":
    # TODO: build an httpx.MockTransport that returns 503 twice then 200,
    #  wrap it as RetryingClient(CircuitBreaker(httpx.AsyncClient(...))),
    #  send one GET and print the status and the attempt count.
    print("Implement the TODOs, then run the tests: pytest 03_api_client_with_retries/")


# ---------------------------------------------------------------------------
# Hints
# ---------------------------------------------------------------------------
# - `httpx.AsyncClient(transport=httpx.MockTransport(handler))` runs the
#   real client stack (headers, response parsing, closing) against a plain
#   function `handler(request) -> httpx.Response`. That's the idiomatic way
#   to fake HTTP in httpx tests -- no monkeypatching, no sockets.
# - Build requests with `client.build_request(...)` or `httpx.Request(...)`
#   and pass them to `send`. A request built with `content=b"..."` can be
#   sent many times; one built from an async generator body can't -- call
#   `await request.aread()` once before the loop to buffer it (and note the
#   memory trade-off for huge uploads).
# - Cancellation needs no special code in asyncio: `await sleep(d)` raises
#   `CancelledError` when the task is cancelled. Your only job is not to
#   swallow it -- `except httpx.TransportError` doesn't catch it, while a
#   careless `except BaseException` would.
# - The breaker needs no lock: asyncio runs one coroutine at a time on a
#   loop, and nothing can interleave between two lines that contain no
#   `await`. Do the check-and-mark ("I am the trial") in one await-free
#   block. If the breaker had to be shared across *threads*, you would need
#   a `threading.Lock` around the same blocks.
#
# Pitfalls
# --------
# - Treating `httpx.TimeoutException` on a POST as safe to retry: the server
#   may have processed it. Idempotency is decided by the request, not by
#   the error type.
# - Retrying inside the breaker instead of outside it. Order matters:
#   RetryingClient(CircuitBreaker(client)) lets the breaker see each
#   attempt and stop the loop with CircuitOpenError; the reverse order
#   counts a whole retry loop as one breaker outcome.
# - Sleeping with `time.sleep` in async code -- it blocks the whole event
#   loop, not just this request.
# - Not resetting the failure counter on success: sporadic failures spread
#   over hours would eventually trip the breaker.
#
# Stretch goals
# -------------
# - A retry *budget* (e.g. retries may be at most 10% of requests over a
#   sliding window), which is what gRPC and Envoy use to cap amplification.
# - One breaker per host (a dict of breakers keyed by `request.url.host`).
# - Emit metrics hooks (attempts, sleeps, state transitions) via a callback.
# - Put the 15_rate_limiter token bucket in front to cap outbound QPS.
